javascript/ql/src/semmle/javascript/security/dataflow/ZipSlip.qll

```diff
@@ -0,0 +1,109 @@
+/**
+ * Provides a taint tracking configuration for reasoning about unsafe zip extraction.
+ */
+
+import javascript
+
+module ZipSlip {
+  /**
+   * A data flow source for unsafe zip extraction.
+   */
+  abstract class Source extends DataFlow::Node { }
+
+  /**
+   * A data flow sink for unsafe zip extraction.
+   */
+  abstract class Sink extends DataFlow::Node { }
+
+  /**
+   * A sanitizer guard for unsafe zip extraction.
+   */
+  abstract class SanitizerGuard extends TaintTracking::SanitizerGuardNode, DataFlow::ValueNode { }
+
+  /** A taint tracking configuration for unsafe zip extraction. */
+  class Configuration extends TaintTracking::Configuration {
+    Configuration() { this = "ZipSlip" }
+
+    override predicate isSource(DataFlow::Node source) { source instanceof Source }
+
+    override predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+
+    override predicate isSanitizerGuard(TaintTracking::SanitizerGuardNode nd) {
+      nd instanceof SanitizerGuard
+    }
+  }
+
+  /**
+   * Gets a node that can be a parsed zip archive.
+   */
+  private DataFlow::SourceNode parsedArchive() {
+    result = DataFlow::moduleImport("unzip").getAMemberCall("Parse")
+    or
+    // `streamProducer.pipe(unzip.Parse())` is a typical (but not
+    // universal) pattern when using nodejs streams, whose return
+    // value is the parsed stream.
+    exists(DataFlow::MethodCallNode pipe |
+      pipe.getMethodName() = "pipe" and
+      parsedArchive().flowsTo(pipe.getArgument(0))
+    )
+  }
+
+  /** A zip archive entry path access, as a source for unsafe zip extraction. */
+  class UnzipEntrySource extends Source {
+    // For example, in
+    // ```javascript
+    // const unzip = require('unzip');
+    //
+    // fs.createReadStream('archive.zip')
+    //   .pipe(unzip.Parse())
+    //   .on('entry', entry => {
+    //      const path = entry.path;
+    //   });
+    // ```
+    // there is an `UnzipEntrySource` node corresponding to
+    // the expression `entry.path`.
+    UnzipEntrySource() {
+      this = parsedArchive()
+            .getAMemberCall("on")
+            .getCallback(1)
+            .getParameter(0)
+            .getAPropertyRead("path")
```

Comment:
Do we want to check somewhere that the `on`-ed event is `"entry"`, or is that unnecessary since we're looking for accesses to `path` anyway?

Reply:
That got lost somewhere during refactoring. I think it's a good idea to have it. Have put it back in.